tests/test_sig_stfl.c

```diff
@@ -0,0 +1,417 @@
+#define PUBLIC_KEY_EXT ".pub"
+
+typedef struct magic_s {
+	uint8_t val[31];
+} magic_t;
+
+void prepend(char *s, const char *t) {
+	size_t len = strlen(t);
+	memmove(s + len, s, strlen(s) + 1);
+	memcpy(s, t, len);
+}
+
+OQS_STATUS lock_sk_key(OQS_SECRET_KEY *sk) {
+	return sk != NULL ? OQS_SUCCESS : OQS_ERROR;
+}
+
+OQS_STATUS release_sk_key(OQS_SECRET_KEY *sk) {
+	return sk != NULL ? OQS_SUCCESS : OQS_ERROR;
+}
+
+OQS_STATUS do_nothing_save(const OQS_SECRET_KEY *sk) {
+	return sk != NULL ? OQS_SUCCESS : OQS_ERROR;
+}
+
+OQS_STATUS sk_file_write(const OQS_SECRET_KEY *sk) {
+	char filename[64];
+	strcpy(filename, (char *)sk->data);
+
+	strcat(filename, PRIVATE_KEY_EXT);
+	FILE *printer = fopen(filename, "w+");
+	if (printer == NULL) {
+		return OQS_ERROR;
+	}
+
+	// Write the entire secret key byte array to the specified file.
+	for (unsigned long i = 0; i < sk->length_secret_key; i++) {
+		if (fputc(sk->secret_key[i], printer) == EOF) {
+			return OQS_ERROR;
+		}
+	}
+	fclose(printer);
+	return OQS_SUCCESS;
+}
+
+/**
+ * @brief Tests the signature generation and verification functions for the
+ * given stateful signature algorithm. Depending on the mode of operation,
+ * the keys are either generated (and saved to a file) or read from a file.
+ *
+ * @param[in] method_name The name of the stateful signature algorithm.
+ * @param[in] mode The mode of operation (reading, saving, nothing).
+ * @param[in] filestem The filestem to use for the key files.
+ * @return OQS_STATUS - OQS_SUCCESS on success, or OQS_ERROR on failure.
+ */
+static OQS_STATUS sig_stfl_test_correctness(const char *method_name, char mode, char *filestem) {
+
+	OQS_SIG_STFL *sig = NULL;
+	uint8_t *public_key = NULL;
+	OQS_SECRET_KEY *secret_key = NULL;
+	uint8_t *message = NULL;
+	size_t message_len = 100;
+	uint8_t *signature = NULL;
+	size_t signature_len;
+	OQS_STATUS rc = OQS_SUCCESS, ret = OQS_ERROR;
+	char filename[64];
+
+	//The magic numbers are random values.
+	//The length of the magic number was chosen to be 31 to break alignment
+	magic_t magic;
+	OQS_randombytes(magic.val, sizeof(magic_t));
+
+	sig = OQS_SIG_STFL_new(method_name);
+	if (sig == NULL) {
+		fprintf(stderr, "ERROR: OQS_SIG_STFL_new failed\n");
+		goto err;
+	}
+
+	printf("================================================================================\n");
+	printf("Sample computation for signature %s\n", sig->method_name);
+	printf("================================================================================\n");
+
+	public_key = malloc(sig->length_public_key + 2 * sizeof(magic_t));
+	message = malloc(message_len + 2 * sizeof(magic_t));
+	signature = malloc(sig->length_signature + 2 * sizeof(magic_t));
+	if ((public_key == NULL) || (message == NULL) || (signature == NULL)) {
+		fprintf(stderr, "ERROR: malloc failed\n");
+		goto err;
+	}
+
+	// Define the secret key object
+	secret_key = OQS_SECRET_KEY_new(method_name);
+	secret_key->data = (void *)filestem;
+	secret_key->lock_key = lock_sk_key;
```

Comment:
Should this be `secret_key->lock_key = &lock_sk_key;` (and similarly for other function pointers in lines below)?  I don't remember precisely how function pointers work...